packages/samples/headless-ssr/src/components/search-page.tsx

```diff
@@ -0,0 +1,22 @@
+'use client';
+
+import {SearchHydrationResult, hydrateInitialState} from '@/src/common/engine';
+import {useEffect, useState} from 'react';
+import ResultList from './result-list';
+
+// TODO: replace "any" with specific type
+export default function SearchPage({engineSnapshot}: {engineSnapshot: any}) {
+  const [hydrationResult, setHydrationResult] =
+    useState<SearchHydrationResult | null>(null);
+
+  useEffect(() => {
+    hydrateInitialState(engineSnapshot).then(({engine, controllers}) => {
+      setHydrationResult({engine, controllers});
+    });
+  }, [engineSnapshot]);
+
+  const {engine} = hydrationResult ?? engineSnapshot;
```

Comment:
One thing I now realize is missing from the RFC, is that we don't have the engine state in the snapshot, only the controller states.

Consequently, I believe `engine` here is `undefined` in `engineSnapshot`, and that's why you can't render anything.

I think that may be fine as-is though, since including the `engine` state in the hydration result would (in Next.js) cause the same issue we discussed in the RFC: we'd send a 75kb+ payload from the server to the client that's mostly unused.

Also, if an implementer needs to obtain a part of the engine state, they can easily build their own `define` controller function which extracts the needed state, and add that to their definition.

Reply:
For now, my recommendation would be to either:
1. Do the workaround I suggested in the comment above. Create a `defineResultList` function inside the sample which just returns `{state: engine.state.search.results}`.
2. Create the `defineResultList` function in `headless/ssr` right away and reject the ticket for it in the backlog.